### PR TITLE
Store specific timezone names by offset, remove `ZoneVariant`

### PR DESCRIPTION
**Pull request by robertbastian**
#5466

### CODE REVIEW

File: provider/source/src/time_zones/convert.rs
Comment (sffc) [inline provider/source/src/time_zones/convert.rs:145]: Question: do you get any warnings printed?

File: provider/source/data/debug/time_zone/specific_long@1/en-001.json
Comment (sffc) [inline provider/source/data/debug/time_zone/specific_long@1/en-001.json:5]: Observation: you are losing some names. There is no "Acre Summer Time" in the new data file.

File: provider/source/src/time_zones/convert.rs
Comment (sffc) [inline provider/source/src/time_zones/convert.rs:135]: Question: do these "sketchy" cases, especially the second one, happen in data?

**Review by sffc — APPROVED** [review]
Praise: this is great. I'm not 100% sure the datagen code is correct, but I think we can clean it up later. As noted in the issue, I think the worst case is we just split some metazones if they have inconsistent offsets.

File: provider/source/src/time_zones/convert.rs
Comment (robertbastian) [reply provider/source/src/time_zones/convert.rs:135]: Yes, for example Ireland has DST in the winter for some reason.

File: provider/source/data/debug/time_zone/specific_long@1/en-001.json
Comment (robertbastian) [reply provider/source/data/debug/time_zone/specific_long@1/en-001.json:5]: Yes, because no time zone in that meta zone actually has a summer time. I think CLDR is inventing some names.

File: provider/source/src/time_zones/convert.rs
Comment (robertbastian) [reply provider/source/src/time_zones/convert.rs:145]: No, only the Minsk thing.

File: provider/source/data/debug/time_zone/specific_long@1/en-001.json
Comment (sffc) [reply provider/source/data/debug/time_zone/specific_long@1/en-001.json:5]: Well it looks like you are only considering the tzdb footer, which has only the "latest" time zone data. It's possible that some of these zones have transitioned away from having a daylight savings transition, but had them previously.

Do I think we should include such data? Well, no, maybe not. It seems like a good way to cut out cold data, and worst cases is they just get the UTC offset printed.

File: provider/source/src/time_zones/convert.rs
Comment (sffc) [reply provider/source/src/time_zones/convert.rs:135]: Ok, are you testing that the Ireland time zone display names work out correctly?

Comment (robertbastian) [reply provider/source/src/time_zones/convert.rs:135]: There's an Ireland test in the fixtures. Before it had to consider the weirdness, now it just looks like any other test.

File: provider/source/data/debug/time_zone/specific_long@1/en-001.json
Comment (robertbastian) [reply provider/source/data/debug/time_zone/specific_long@1/en-001.json:5]: I think it makes sense to include only what is valid today. If we apply today's tz names to a date from 1980, people will understand. If we include tz names that haven't been used for 40 years, or worse, meant something different 40 years ago, that might be technically correct, but non tz-nerds will not interpret it correctly.

File: provider/source/src/time_zones/convert.rs
Comment (robertbastian) [reply provider/source/src/time_zones/convert.rs:149]: These are the warnings when I don't allow DST divergence: 
```
"Africa/Accra" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Bamako" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Banjul" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Conakry" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Dakar" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Freetown" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Lome" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Nouakchott" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
"Africa/Tunis" has different offsets from "Europe/Paris": standard 1 vs 1, daylight None vs Some(2.0)
"America/Anguilla" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Antigua" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Aruba" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Barbados" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Belize" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Cayman" has different offsets from "America/New_York": standard -5 vs -5, daylight None vs Some(-4.0)
"America/Costa_Rica" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Curacao" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Dominica" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/El_Salvador" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Grenada" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Guadeloupe" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Guatemala" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Jamaica" has different offsets from "America/New_York": standard -5 vs -5, daylight None vs Some(-4.0)
"America/Kralendijk" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Lower_Princes" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Marigot" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Martinique" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Mexico_City" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Montserrat" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Panama" has different offsets from "America/New_York": standard -5 vs -5, daylight None vs Some(-4.0)
"America/Port_of_Spain" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Puerto_Rico" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/St_Kitts" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/St_Lucia" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/St_Thomas" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/St_Vincent" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"America/Tegucigalpa" has different offsets from "America/Chicago": standard -6 vs -6, daylight None vs Some(-5.0)
"America/Tortola" has different offsets from "America/Halifax": standard -4 vs -4, daylight None vs Some(-3.0)
"Atlantic/St_Helena" has different offsets from "Atlantic/Reykjavik": standard 0 vs 0, daylight None vs Some(1.0)
```

It appears that metazones don't imply DST equality.

Comment (sffc) [reply provider/source/src/time_zones/convert.rs:149]: I think those error messages all are of the form "standard x vs x, daylight y vs None", yes? So that just means there are some countries sharing the same metazone but not all observe DST. It should still be safe to add the display name.

**Comment by robertbastian** [comment]
I've updated the TZ logic to not use the POSIX string, but to load the TZDB from Github and to actually look at the most recent historical DST. This way, display names for past DSTs that are not observed anymore are included. The errors that this approach surfaces are:

* The metazone `samo` (Samoa) has a DST display name, but it has never used DST
* The metazones `anad`, `aqta`, `arwe`, `cave`, `falk`, `geor`, `kamc`, `kyrg`, `novo`, `sakh`, `sama`, `turk`, `yuko` are currently using an offset which corresponds to the last DST offset they have used. We can therefore not store the historical DST name. However, I think doing so would be confusing anyway.

File: provider/source/Cargo.toml
Comment (sffc) [inline provider/source/Cargo.toml:59]: Question: why did you switch from `jiff_tzdb` + `tzif` to `parse-zoneinfo`?

We know the maintainers of the first two crates, and of course `tzif` lives in this repository.

File: provider/source/data/tzdb/africa
Comment (sffc) [inline provider/source/data/tzdb/africa:1]: Question: why did you add this file directly? We can load the text files from a bunch of different places. In the repo, they should live next to the debug source CLDR data.

File: provider/source/src/lib.rs
Comment (sffc) [inline provider/source/src/lib.rs:230]: Nit: Please write the docs now while this is fresh. It takes 60 seconds. (In general, not just this PR)

File: provider/source/src/time_zones/convert.rs
Comment (sffc) [inline provider/source/src/time_zones/convert.rs:117]: Issues/Suggestions:

1. `map_zone.zone_type` gives you only a single "golden zone" for the given metazone and territory. There might be more IANA zones in the same territory that you are missing. For example, I don't know if you ever reach a zone like America/Phoenix with this code.
2. IANA zones can switch between metazones, and you don't account for that here. You can see in the data when a zone switches metazones. You should ideally read the TZDB entries only for the period in which the zone was in the metazone.
3. I think a better approach might be to flip this around: start with the IANA zone, loop over all metazones, load the offsets for the given time period, and then insert them into a map from metazones to offsets. By the end, you will have comprehensively touched all time zones and metazones.

File: provider/source/Cargo.toml
Comment (robertbastian) [reply provider/source/Cargo.toml:59]: The TZDB is distributed in text format at https://github.com/eggert/tz. It is not distributed anywhere we could find in binary format (#). Using `jiff_tzdb` means Burntsushi has to update their crate every time a new tzdb release happens, and users then have to recompile `icu4x-datagen`.

File: provider/source/data/tzdb/africa
Comment (robertbastian) [reply provider/source/data/tzdb/africa:1]: I'm planning to ship these. We're also shipping segmentation data.

Comment (sffc) [reply provider/source/data/tzdb/africa:1]: We only ship the segmentation data because Makoto hand-wrote the rules; the long-term plan is to upstream the hand-written rules into CLDR/Unicode.

I'm surprised that you are the one suggesting that we ship these time zone files in datagen.

Comment (robertbastian) [reply provider/source/data/tzdb/africa:1]: I'm leaning towards shipping as I want to hide this from clients. I don't want to give the impression that ICU4X data has to be regenerated with each tzdb update.

Comment (sffc) [reply provider/source/data/tzdb/africa:1]: I'm not opposed to that way of thinking, but then we maybe can use `jiff-tzdb`?

**Comment by sffc** [comment]
I see two main actions here:

1. Address the concern I raised involving metazones that are valid for certain ranges
2. Do the homework from @macchiati of making a few example cases where the ICU4X behavior after landing this change will differ from ICU4C behavior.

File: components/datetime/src/provider/time_zones.rs
Comment (nekevss) [inline components/datetime/src/provider/time_zones.rs:139]: Is this suppose to be `EighthsOfHourOffset`?

File: utils/tzif/src/lib.rs
Comment (nekevss) [inline utils/tzif/src/lib.rs:42]: Definitely think this is better. I ran into not being able to read bytes and though it was weird. But if it's a change to main functionality. Would it be worth having an example for how to read from a file? In case someone is using `tzif` elsewhere.

**Review by nekevss — COMMENTED** [review]
Ooooh, nice! I was looking through and had a couple questions / things I noticed.

File: provider/source/Cargo.toml
Comment (nekevss) [reply provider/source/Cargo.toml:59]: The files can also be found at https://data.iana.org/time-zones/. But to get the binary formats `zic.c` would need to be implemented in Rust or some other alternative.

File: provider/source/data/tzdb/africa
Comment (nekevss) [reply provider/source/data/tzdb/africa:1]: Another possible option might be to lazy download the files from one of the couple URLs they're available from (thinking about the GitHub mirror or IANA's data URL), and maybe add a TZDB CLI argument to provide either a directory of zoneinfo files to be compiled or a directory of precompiled TZifs.

Comment (sffc) [reply provider/source/data/tzdb/africa:1]: We default to `new_latest_tested` for CLDR data. Seems like we can download clean TZDB files in the same way? We would not pull from main unless the user specifically requests that.

Comment (robertbastian) [reply provider/source/data/tzdb/africa:1]: > I'm not opposed to that way of thinking, but then we maybe can use jiff-tzdb?

`jiff-tzdb` is a set of embedded zic files, it's not a parser.

I can make it auto-download.

Comment (robertbastian) [reply provider/source/data/tzdb/africa]: Circling back on this, the text sources are actually more powerful than the compiled tzifs. Afaict the tzifs contain only transitions defined in unix timestamps, whereas the source files contain more structure, they call it zoneinfo (bascially what the standard time is for a period), and rules (when dst is observed, and what the offset is). The tzif transitions alone are not sufficient to compute our offset periods.

**Pull request closed without merging**
